spec/app/ics-030-middleware/BEST_PRACTICES.md

```diff
@@ -0,0 +1,71 @@
+# Best Practices When Designing Middleware
+
+## Namespacing the Stack
+
+Each layer of the middleware should reserve a unique name in the stack so that it can key into specific information it needs in the version, packet data, or acknowledgement. Since this will not be in the portID which is shared across the stack, it will not be enforced by IBC. The Cosmos-SDK, for example, can accomplish uniqueness by using the ModuleName to reference the different middleware in the stack.
+
+## Version Negotiation For Middleware
+
+Middleware will only need to negotiate versions if they expect the counterparty chain to have compatible middleware on the stack as well. Middleware that can execute unilaterally **do not** need version negotiation.
+
+Middleware must take care to ensure that the application logic can execute its own version negotiation without interference from the nesting middleware. In order to do this, the middleware will format the version in a JSON-encoded string. Any information intended for the middleware will be keyed on the middleware namespace. The version intended for the underlying app will be keyed on `"app_version"`. The application version may as well be a JSON-encoded string, possibly including further middleware and app versions, if the application stack consists of multiple milddlewares wrapping a base application. The version values themselves may be more complex than static strings, they can be JSON encoded structs with multiple fields that the counterparties must agree on.  The format of the version string is as follows:
+
+```json
+{
+    "<middleware_name>": "<middleware_version_value>",
+    "app_version": "<application_version_value>",
+}
+```
```

Comment:
Is there an issue tracking the work of making the version/packet data/ack a flat map? I think that improvement would be a really nice dev UX improvement